File: nodejs/lame.scanner.js
'use strict';

var os = require('os');
var fs = require('fs');
var path = require('path');
var Lame = require('./lame.js');

////////////////////////////////////////////////////////////////
var lame = null;

function Scanner() {
	try {
		lame = new Lame();
	} catch (e) {
		console.log(e);
	}
}

function isDirectory(filepath) {
	try {
		var info = fs.statSync(filepath);
		return info.isDirectory();
	} catch (e) {
		console.log(e);
	}
}

function DealFile(filename) {
	try {
		var dr = lame.ScanFile(filename);
		if (null != dr) {
			console.warn(filename + '\tInfected: ' + dr.vname + '(' + dr.engid + ')');
		} else {
			console.log(filename);
		}
	} catch (e) {
		console.log(e);
	}
}

function DealFileWithCallback(filename, enter_file, leave_file, alarm, userdata) {
	try {
		return lame.ScanFileWithCallback(filename, enter_file, leave_file, alarm, userdata);
	} catch (e) {
		console.log(e);
	}
	return false;
}

function readDirSync(filepath, list) {
	try {
		var pa = fs.readdirSync(filepath);
		pa.forEach(function (filename, index) {
			var fp = filepath + "/" + filename;
			if (isDirectory(fp)) {
				readDirSync(fp, list);
			} else {
				list.push(fp);
			}
		});
	} catch (e) {
		console.log(e);
	}
}

function GetFileList(filepath) {
	var list = [];
	try {
		if (isDirectory(filepath)) {
			readDirSync(filepath, list);
		} else {
			list.push(filepath);
		}
	} catch (e) {
		console.log(e);
	}
	return list;
}

Scanner.prototype.Load = function Load() {
	try {
		if (!lame.OpenVdb(null)) {
			throw ('lame.OpenVdb failed.');
		}
		if (!lame.Load()) {
			throw ('lame.Load failed.');
		}
		return true;
	} catch (e) {
		console.log(e);
		lame.Unload();
	}
	return false;
}

Scanner.prototype.Unload = function Unload() {
	try {
		lame.CloseVdb();
		lame.Unload();
	} catch (e) {
		console.log(e);
	}
}

Scanner.prototype.ScanFile = function ScanFile(filepath) {
	if (null == filepath) return;
	try {
		var list = GetFileList(filepath);
		while (list.length > 0) {
			var filename = list.shift();
			DealFile(filename);
		}
	} catch (e) {
		console.log(e);
	}
}

Scanner.prototype.ScanFileWithCallback = function ScanFileWithCallback(filepath, enter_file, leave_file, alarm, userdata) {
	if (null == filepath) return;
	try {
		var list = GetFileList(filepath);
		for (var key in list) {
			if (list.hasOwnProperty(key)) {
				var element = list[key];
				DealFileWithCallback(element, enter_file, leave_file, alarm, userdata);
			}
		}
	} catch (e) {
		console.log(e);
	}
}

Scanner.prototype.ScanMem = function ScanMem(data, size) {
	if (null == data || 0 == size) return;
	try {
		var dr = lame.ScanMem(data, size);
		if (null != dr) {
			console.warn(data + '\tInfected: ' + dr.vname + '(' + dr.engid + ')');
		} else {
			console.log(data);
		}
	} catch (e) {
		console.log(e);
	}
}

Scanner.prototype.ScanMemWithCallback = function ScanMemWithCallback(data, size, enter_file, leave_file, alarm, userdata) {
	try {
		return lame.ScanMemWithCallback(data, size, enter_file, leave_file, alarm, userdata);
	} catch (e) {
		console.log(e);
	}
	return false;
}

Scanner.prototype.ShowVersion = function ShowVersion() {
	try {
		var version = lame.GetVersionInfo();
		if (null != version) {
			console.log('engine version: ' + version.engv);
			console.log('viruslib version: ' + version.vdbv);
		} else {
			console.log('GetLameInfo failed.');
		}
	} catch (e) {
		console.log(e);
	}
}

Scanner.prototype.ShowLicence = function ShowLicence() {
	try {
		var info = lame.GetLicenceInfo();
		if (null != info) {
			console.log('Version: ' + info.Version);
			console.log('Owner: ' + info.Owner);
			console.log('Date: ' + info.Date);
			console.log('Authm: ' + info.Authm);
			console.log('Data: ' + info.Data);
		} else {
			console.log('GetLicenceInfo failed.');
		}
	} catch (e) {
		console.log(e);
	}
}

Scanner.prototype.ExtractFile = function ExtractFile(filename, password, enter_file, leave_file, userdata) {
	try {
		return lame.ExtractFile(filename, password, enter_file, leave_file, userdata);
	} catch (e) {
		console.log(e);
	}
	return false;
}

Scanner.prototype.ExtGetSize = function ExtGetSize(handle) {
	try {
		if (null != handle) {
			return lame.Extract_GetSize(handle);
		}
	} catch (e) {
		console.log(e);
	}
	return 0;
}

Scanner.prototype.ExtSeek = function ExtSeek(handle, offset, method) {
	try {
		return lame.Extract_Seek(handle, offset, method);
	} catch (e) {
		console.log(e);
	}
	return false;
}

Scanner.prototype.ExtTell = function ExtTell(handle) {
	try {
		return lame.Extract_Tell(handle);
	} catch (e) {
		console.log(e);
	}
	return -1;
}

Scanner.prototype.ExtRead = function ExtRead(handle, size) {
	try {
		if (null != handle && size > 0) {
			return lame.Extract_Read(handle, size);
		}
	} catch (e) {
		console.log(e);
	}
	return null;
}

var LAME_SEEK_SET	= 0;
var LMAE_SEEK_CUT	= 1;
var LMAE_SEEK_TAIL	= 2;

Scanner.prototype.ExtReadPos = function ExtReadPos(handle, offset, size) {
	try {
		if (null != handle && offset >= 0 && size > 0) {
			if (!lame.Extract_Seek(handle, offset, LAME_SEEK_SET))
				return null;
			return lame.Extract_Read(handle, size);
		}
	} catch (e) {
		console.log(e);
	}
	return null;
}

module.exports = Scanner;

File: python/lame.scanner.py
import os
import sys
import platform
import signal
import ccolor
from lame import *

def format_bstring(bstr, CodePage=None):
    if 3 == sys.version_info.major:
        if CodePage is None:
            return bstr.decode()
        else:
            return bstr.decode(CodePage)
    else:
        return bstr

class LameScanner:

    def __init__(self, lpath):
        self._lame = Lame(lpath)
        self._color = ccolor.ConsoleColor()

    def Load(self, vdb):
        if self._lame is None:
            return
        if not self._lame.Load(vdb):
            return
        return True

    def SetParam(self, param):
        if param is None:
            return
        if self._lame is None:
            return
        self._lame.SetParam(param)
        return


    def Scan(self, path_):
        if os.path.isfile(path_):
            self.__ScanFile(path_)
        else:
            self.__ScanDir(path_)
        return

    def Unload(self):
        if self._lame is not None:
            self._lame.Unload()
            self._lame = None
        return

    def ShowLicense(self):
        if self._lame is None:
            return
        _license = self._lame.GetLicense()
        if _license is None:
            print("get licence information failed.")
            return
        print("License:")
        print("     Version: " + format_bstring(_license.version))
        print("     Owner  : " + format_bstring(_license.Owner, 'gb2312'))
        print("     Date   : " + format_bstring(_license.Data))
        print("     Authm  : " + format_bstring(_license.Authm))
        print("     Data   : " + format_bstring(_license.Data) + "\n")
        return

    def ShowVersion(self):
        if self._lame is None:
            return
        _version = self._lame.GetVersion()
        if _version is None:
            return
        print("Engine Version    :" + format_bstring(_version.engv))
        print("Viruse Lib Version:" + format_bstring(_version.vdbv))
        return

    def __ScanFile(self, fname):
        if not os.path.exists(fname):
            return
        if self._lame is None:
            return
        sys.stdout.write(fname)
        _result = self._lame.ScanFile(fname)
        if _result is not None:
            sys.stdout.write(" " + "  Infected: " + format_bstring(_result.vname) + " (" + format_bstring(_result.engid) + ")")
        sys.stdout.write("\n")
        return

    def __ScanDir(self, dir_):
        if not os.path.exists(dir_):
            return
        for dirpath,dirnames,filenames in os.walk(dir_):
            for file in filenames:
                fullpath=os.path.join(dirpath,file)
                self.__ScanFile(fullpath)
        return


class LameScanner2:

    def __init__(self, lpath):
        self._lame = LameWithFeedback(lpath)
        self._lame.SetCallack(self.__EnterFile, self.__LeaveFile, self.__Alarm)
        self._color = ccolor.ConsoleColor()
        self._virus_info_list = []

    def Load(self, vdb):
        if self._lame is None:
            return
        if not self._lame.Load(vdb):
            return
        return True

    def SetParam(self, param):
        if param is None:
            return
        if self._lame is None:
            return
        self._lame.SetParam(param)
        return

    def Unload(self):
        if self._lame is not None:
            self._lame.Unload()
            self._lame = None
        return

    def ShowLicense(self):
        if self._lame is None:
            return
        _license = self._lame.GetLicense()
        if _license is None:
            print("get licence information failed.")
            return
        print("License:")
        print("     Version: " + format_bstring(_license.version))
        print("     Owner  : " + format_bstring(_license.Owner))
        print("     Date   : " + format_bstring(_license.Data))
        print("     Authm  : " + format_bstring(_license.Authm))
        return

    def ShowVersion(self):
        if self._lame is None:
            return
        _version = self._lame.GetVersion()
        if _version is None:
            return
        print("Engine Version    : " + format_bstring(_version.engv))
        print("Viruse Lib Version: " + format_bstring(_version.vdbv))
        return

    def __EnterFile(self, fname, depth):
        self._virus_info_list.append([fname, ""])
        return LameWithFeedback.LSCT_CONTINUE

    def __LeaveFile(self, fname, depth):
        if self._virus_info_list.count == 0:
            return
        vinfo = self._virus_info_list[len(self._virus_info_list) - 1]
        self._virus_info_list.remove(vinfo)
        sys.stdout.write(vinfo[0])
        if len(vinfo[1]) != 0:
            sys.stdout.write(" " + "  Infected: " + vinfo[1])
        sys.stdout.write("\n")
        return

    def __Alarm(self, fname, result):
        if result is None:
            return LameWithFeedback.LSCT_CONTINUE
        self._virus_info_list[len(self._virus_info_list) - 1][1] = result.vname + " (" + result.engid+")"
        return LameWithFeedback.LSCT_CONTINUE

    def Scan(self, path_):
        if os.path.isfile(path_):
            self.__ScanFile(path_)
        else:
            self.__ScanDir(path_)
        return

    def __ScanFile(self, fname):
        if not os.path.exists(fname):
            return
        if self._lame is None:
            return
        self._lame.ScanFile(fname)
        return

    def __ScanDir(self, dir_):
        if not os.path.exists(dir_):
            return
        for dirpath,dirnames,filenames in os.walk(dir_):
            for file in filenames:
                fullpath=os.path.join(dirpath,file)
                self.__ScanFile(fullpath)
        return


def exit(signum, frame):
    sys.exit()

def main(lame_path, argv):
    _lame_path = lame_path
    _param_lst = []
    _scan_path_lst = []

    bShowVersion = False
    bShowLicense = False
    bShowFileList = False

    for _param in argv:
        if _param.startswith('-'):
            if _param.lower() == "-version":
                bShowVersion = True
            elif _param.lower() == "-license":
                bShowLicense = True
            elif _param.lower() == "-show-file-list":
                bShowFileList = True
            else:
                _param_lst.append(_param[1:])
        else:
            _scan_path_lst.append(_param)

    _lame = None
    if bShowFileList:
        _lame = LameScanner2(_lame_path)
    else:
        _lame = LameScanner(_lame_path)

    if bShowLicense:
        _lame.ShowLicense()

    if bShowVersion:
        _lame.ShowVersion()

    if 0 == len(_scan_path_lst):
        return

    _dbf = VirusDb(_lame_path)
    if not _dbf.OpenVdb(None):
        print("Faild to openvdb.")
        return

    for _param in _param_lst:
        _lame.SetParam(_param)

    if not _lame.Load(_dbf):
        _dbf.CloseVdb()
        print("Faild to load lame.")
        return

    for _path in _scan_path_lst:
        _lame.Scan(_path)

    _lame.Unload()
    _dbf.CloseVdb()
    return

# sys.argv.append('D:\\MyJob\\SDK\\produce\\make\\lame-win-x64')
# sys.argv.append('D:\\rsdata\\1\\11')

if  __name__ == "__main__":
    signal.signal(signal.SIGINT, exit)
    signal.signal(signal.SIGTERM, exit)
    main(sys.argv[1], sys.argv[2:])
